pydantic_ai_slim/pydantic_ai/toolsets/_dynamic.py

```diff
@@ -17,39 +16,92 @@
 
 
-@dataclass
+class ToolsetRunStep(TypedDict, Generic[AgentDepsT]):
+    """State for a DynamicToolset for a specific run."""
+
+    toolset: AbstractToolset[AgentDepsT] | None
+    run_step: int | None
+
+
 class DynamicToolset(AbstractToolset[AgentDepsT]):
     """A toolset that dynamically builds a toolset using a function that takes the run context.
 
-    It should only be used during a single agent run as it stores the generated toolset.
-    To use it multiple times, copy it using `dataclasses.replace`.
+    State is isolated per run using `ctx.run_id` as a key, allowing the same instance
+    to be safely reused across multiple agent runs.
     """
 
     toolset_func: ToolsetFunc[AgentDepsT]
-    per_run_step: bool = True
-
-    _toolset: AbstractToolset[AgentDepsT] | None = None
-    _run_step: int | None = None
+    per_run_step: bool
+    _id: str | None
+    _toolset_runstep: dict[str, ToolsetRunStep[AgentDepsT]]
+
+    def __init__(
+        self,
+        toolset_func: ToolsetFunc[AgentDepsT],
+        *,
+        per_run_step: bool = True,
+        id: str | None = None,
+    ):
+        """Build a new dynamic toolset.
+
+        Args:
+            toolset_func: A function that takes the run context and returns a toolset or None.
+            per_run_step: Whether to re-evaluate the toolset for each run step. Defaults to True.
+            id: An optional unique ID for the toolset. A toolset needs to have an ID in order to be used
+                in a durable execution environment like Temporal, in which case the ID will be used to
+                identify the toolset's activities within the workflow.
+        """
+        self.toolset_func = toolset_func
+        self.per_run_step = per_run_step
+        self._id = id
+        self._toolset_runstep = {}
 
     @property
     def id(self) -> str | None:
-        return None  # pragma: no cover
+        return self._id
+
+    def copy(self) -> DynamicToolset[AgentDepsT]:
+        """Create a copy of this toolset for use in a new agent run."""
+        return DynamicToolset(
+            self.toolset_func,
+            per_run_step=self.per_run_step,
+            id=self._id,
+        )
 
     async def __aenter__(self) -> Self:
         return self
 
     async def __aexit__(self, *args: Any) -> bool | None:
         try:
-            if self._toolset is not None:
-                return await self._toolset.__aexit__(*args)
+            result = None
+            for run_state in self._toolset_runstep.values():
+                if run_state['toolset'] is not None:
+                    result = await run_state['toolset'].__aexit__(*args)
         finally:
-            self._toolset = None
-            self._run_step = None
+            self._toolset_runstep.clear()
+        return result
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, DynamicToolset):
+            return False
```

Comment:
It reads a little easier to ad `isinstance(other, DynamicToolset) and` to the next lines